pkg/ccrm/ccrm_handler.go

```diff
@@ -55,29 +76,112 @@ type NodeAttributesData struct {
+	s.cloudletSSHKey = cloudletssh.NewSSHKey(s.vaultClient)
+	s.crmHandler = crmutil.NewCRMHandler(s.getCRMCloudletPlatform, s.nodeMgr)
+	s.proxyCertsCache = certscache.NewProxyCertsCache(s.vaultClient)
 
-	// notify handlers
-	s.caches.CloudletCache.AddUpdatedCb(s.cloudletChanged)
 	s.caches.CloudletNodeCache.AddUpdatedCb(s.cloudletNodeChanged)
+	s.crmHandler.SettingsCache.AddUpdatedCb(s.crmHandler.SettingsChanged)
+}
 
-	if redisClient != nil {
-		// redis handlers
-		hctx, cancel := context.WithCancel(ctx)
-		s.CancelHandlers = cancel
-		server := rediscache.GetCCRMAPIServer(redisClient, nodeType, s)
-		server.Start(hctx)
+func (s *CCRMHandler) InitConnectivity(client *notify.Client, kvstore objstore.KVStore, nodeMgr *node.NodeMgr, grpcServer *grpc.Server, sync edgeproto.DataSync) {
+	// Caches are updated via etcd watches.
+	// In general most data sent from here to the Controller is
+	// via return values from GRPC API calls the Controller makes to here.
+	s.sync = sync
+	s.crmHandler.AppCache.InitSync(sync)
+	s.crmHandler.AppInstCache.InitSync(sync)
+	s.crmHandler.CloudletInternalCache.InitSync(sync)
+	s.crmHandler.CloudletCache.InitSync(sync)
+	s.crmHandler.VMPoolCache.InitSync(sync)
+	s.crmHandler.FlavorCache.InitSync(sync)
+	s.crmHandler.ClusterInstCache.InitSync(sync)
+	s.crmHandler.TrustPolicyCache.InitSync(sync)
+	s.crmHandler.TrustPolicyExceptionCache.InitSync(sync)
+	s.crmHandler.AutoProvPolicyCache.InitSync(sync)
+	s.crmHandler.AutoScalePolicyCache.InitSync(sync)
+	s.crmHandler.SettingsCache.InitSync(sync)
+	s.crmHandler.ResTagTableCache.InitSync(sync)
+	s.crmHandler.GPUDriverCache.InitSync(sync)
+	s.crmHandler.AlertPolicyCache.InitSync(sync)
+	s.crmHandler.NetworkCache.InitSync(sync)
+	s.caches.PlatformFeaturesCache.InitSync(sync)
+	s.caches.CloudletNodeCache.InitSync(sync)
+	s.caches.CloudletInfoCache.InitSync(sync)
+	s.caches.ClusterInstInfoCache.InitSync(sync)
+	s.caches.AppInstInfoCache.InitSync(sync)
+	nodeMgr.CloudletLookup.GetCloudletCache(node.NoRegion).InitSync(sync)
+	nodeMgr.CloudletPoolLookup.GetCloudletPoolCache(node.NoRegion).InitSync(sync)
+
+	// notify handlers
+	if client != nil {
+		client.RegisterSendAlertCache(&s.crmHandler.AlertCache)
+		client.RegisterRecv(notify.NewFedAppInstEventRecv(s))
+		nodeMgr.RegisterClient(client)
 	}
-	s.vaultClient = accessapi.NewVaultClient(ctx, nodeMgr.VaultConfig, s, flags.Region, flags.DnsZone, nodeMgr.ValidDomains)
+
+	// grpc handlers
+	if grpcServer != nil {
+		edgeproto.RegisterCloudletPlatformAPIServer(grpcServer, s)
+		edgeproto.RegisterClusterPlatformAPIServer(grpcServer, s)
+		edgeproto.RegisterAppInstPlatformAPIServer(grpcServer, s)
+	}
+}
+
+func (s *CCRMHandler) Start(ctx context.Context, ctrlConn *grpc.ClientConn) {
+	s.ctrlConn = ctrlConn
+	for _, builder := range s.platformBuilders {
+		plat := builder()
+		features := plat.GetFeatures()
+		features.NodeType = s.nodeMgr.MyNode.Key.Type
```

Comment:
`fatures.NodeType = s.nodeType`
otherwise, what's the point of an extra member field in CCRMHandler :)

Reply:
Good point :)